walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,121 @@
 
 package com.google.sps;
 
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Arrays;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+
+    private final ArrayList < TimeRange > options = new ArrayList < TimeRange > ();
+
+    public Collection < TimeRange > query(Collection < Event > events, MeetingRequest request) {
+
+        Collection attendees = request.getAttendees();
+        Collection optionalAttendees = request.getOptionalAttendees();
+
+        // check if duration is longer than the day. if so, return empty collection.
+        long duration = request.getDuration();
+        if (duration > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+
+        // if no attendees or optional attendees return collection of one item being the full day
+        if (attendees.isEmpty() && optionalAttendees.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+
+        // if there are optional attendees but no attendees, optional attendees will be treated as attendees
+        } else if (attendees.isEmpty()) {
+            attendees = optionalAttendees;
+        }
+
+        // if no events return the whole day
+        ArrayList < Event > eventsList = new ArrayList < Event > (events);
+        if (eventsList.isEmpty() == true) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        // Sorts array of event objects by start time in ascending order
+        Collections.sort(eventsList, Event.ORDER_BY_START_TIME);
+        Iterator < Event > eventsListIterator = eventsList.iterator();
+
+        // Conflicting event is found
+        Event conflict = eventsListIterator.next();
+        int start;
+
+        // Ignore conflict if it doesn't involve any person in the request 
+        if (conflictIsRelevant(conflict, attendees) == false) {
+            start = TimeRange.START_OF_DAY;
+        } else {
+
+            // add duration between start of day and first conflict
+            int possibleDurationStart = TimeRange.START_OF_DAY;
+            int possibleDurationEnd = conflict.getWhen().start();
+            if (slotDurationIsPossible(request, possibleDurationStart, possibleDurationEnd)) {
+                options.add(TimeRange.fromStartEnd(possibleDurationStart, possibleDurationEnd, false));
+            }
+
+            // start begins at the end of first conflict
+            start = conflict.getWhen().end();
+        }
+
+        while (eventsListIterator.hasNext()) {
```

Comment:
doing work outside a loop that's the same as inside the loop work is generally a code smell and an indication that there might be a simpler way to accomplish things. Can you get rid of the work you are doing before this loop that is the same as the work inside the loop?

Reply:
Previously, I implemented it that way because I had to initialize conflict outside the loop in order to get prevConflict. That's useful in the case of overlapping events. I sorted events based on start time, so for example event[1] starts 8:30 ends 10:00 and event[2] starts 9:00 and ends 9:30, so I have to switch conflict back to event one.

I also had issues with the eventSplitRestriction  test i.e  (only one conflict). 
If I initialize "conflict = eventsListIterator.next();" outside the loop, then I can't get into the while-loop because there is no next conflict. So, I had to solve the startOfDday to firstConflict case outside the loop.

I was able to solve the problems by initializing conflict and prevConflict to null before getting into the loop. I'm not sure if that's good code style?

Reply:
In this case I would say that is the correct decision. It allows you to remove redundant code, and simplify your other code. Generally, speaking, the less code means less places for bugs to hide, and less for a reader to understand.